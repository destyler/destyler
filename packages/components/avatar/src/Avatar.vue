<script setup lang="ts">
import { onMounted, ref } from 'vue'
import type { DestylerAvatarProps } from './props'

defineOptions({
  name: 'DestylerAvatar',
})

const { src, alt, delayMs } = withDefaults(defineProps<DestylerAvatarProps>(), {
  delayMs: 600,
})

const canRender = ref(false)
let timeout: ReturnType<typeof setTimeout> | undefined

if (delayMs) {
  timeout = setTimeout(() => {
    canRender.value = true
    clearTimeout(timeout)
  }, delayMs)
}
else {
  canRender.value = true
}

onMounted(() => {
  imageInit(src!)
})

const imageLoadingStatus = ref<'loaded' | 'error'>()
function imageInit(src: string) {
  const image = new Image()
  image.src = src
  image.onload = () => {
    imageLoadingStatus.value = 'loaded'
  }
  image.onerror = () => {
    imageLoadingStatus.value = 'error'
  }
}
</script>

<template>
  <span
    destyler-avatar-root
  >
    <img
      v-if="imageLoadingStatus === 'loaded'"
      destyler-avatar-img
      :src="src"
      :alt="alt"
    >
    <span
      v-if="imageLoadingStatus !== 'loaded' && canRender"
      destyler-avatar-fallback
    >
      {{ fallback }}
    </span>
  </span>
</template>
