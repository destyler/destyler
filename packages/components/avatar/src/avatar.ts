import type { ImgHTMLAttributes, PropType, VNodeChild } from 'vue'
import { defineComponent, h, onBeforeUnmount, onMounted, ref, watch, watchEffect } from 'vue'
import { isImageSupportNativeLazy, observeIntersection, resolveSlot, resolveWrappedSlot } from '@destyler/shared'
import { DestylerResizeObserver } from '@destyler/resize-observer'

export type IntersectionObserverOptions = Omit<
IntersectionObserverInit,
'root'
> & {
  root?: Element | Document | null | string
}

const DestylerAvatar = defineComponent({
  name: 'DestylerAvatar',
  props: {
    src: {
      type: String as PropType<string>,
    },
    fallbackSrc: {
      type: String as PropType<string>,
    },
    imgProps: {
      type: Object as PropType<ImgHTMLAttributes>,
    },
    intersectionObserverOptions: {
      type: Object as PropType<IntersectionObserverOptions>,
    },
    lazy: {
      type: Boolean as PropType<boolean>,
    },
    renderFallback: {
      type: Function as PropType<() => VNodeChild>,
    },
    renderPlaceholder: {
      type: Function as PropType<() => VNodeChild>,
    },
    onError: {
      type: Function as PropType<(e: Event) => void>,
    },
    onLoad: {
      type: Function as PropType<(e: Event) => void>,
    },
  },
  setup(props, { slots }) {
    let memoedTextHtml: string | null = null
    const textRef = ref<HTMLElement | null>(null)
    const selfRef = ref<HTMLElement | null>(null)
    const loaded = ref(!props.lazy)
    const shouldStartLoading = ref(!props.lazy)
    const hasLoadError = ref(false)

    function fitTextTransform(): void {
      const { value: textEl } = textRef
      if (textEl) {
        if (memoedTextHtml === null || memoedTextHtml !== textEl.innerHTML) {
          memoedTextHtml = textEl.innerHTML
          const { value: selfEl } = selfRef
          if (selfEl) {
            const { offsetWidth: elWidth, offsetHeight: elHeight } = selfEl
            const { offsetWidth: textWidth, offsetHeight: textHeight } = textEl
            const radix = 0.9
            const ratio = Math.min(
              (elWidth / textWidth) * radix,
              (elHeight / textHeight) * radix,
              1,
            )
            textEl.style.transform = `translateX(-50%) translateY(-50%) scale(${ratio})`
          }
        }
      }
    }

    function mergedOnLoad(e: Event) {
      props.onLoad?.(e)
      props.imgProps?.onLoad?.(e)
      loaded.value = true
    }

    function handleError(e: Event): void {
      if (!shouldStartLoading.value)
        return
      hasLoadError.value = true
      props.imgProps?.onError?.(e)
      if (props.onError)
        props.onError(e)
    }

    watch(
      () => props.src,
      () => {
        hasLoadError.value = false
      },
    )

    onMounted(() => {
      if (isImageSupportNativeLazy)
        return

      let unobserve: (() => void) | undefined
      const stopWatchHandle = watchEffect(() => {
        unobserve?.()
        unobserve = undefined
        if (props.lazy) {
          unobserve = observeIntersection(
            selfRef.value,
            props.intersectionObserverOptions,
            shouldStartLoading,
          )
        }
      })
      onBeforeUnmount(() => {
        stopWatchHandle()
        unobserve?.()
      })
    })

    return {
      textRef,
      selfRef,
      fitTextTransform,
      hasLoadError,
      handleError,
      shouldStartLoading,
      loaded,
      mergedOnLoad,
    }
  },
  render() {
    let img: VNodeChild
    const placeholderNode = !this.loaded
    && !this.hasLoadError
    && (this.$props.renderPlaceholder
      ? this.$props.renderPlaceholder()
      : this.$slots.placeholder?.())

    if (this.hasLoadError) {
      img = this.$props.renderPlaceholder
        ? this.$props.renderPlaceholder()
        : resolveSlot(this.$slots.fallback, () => [
          h('img', {
            destyler: 'avatar-fallback',
            src: this.$props.fallbackSrc,
          }),
        ])
    }
    else {
      img = resolveWrappedSlot(this.$slots.default, (children) => {
        if (children) {
          const avatatText = h('span', {
            ref: 'textRef',
            destyler: 'avatar-text',
          }, children)
          return h(DestylerResizeObserver, {
            onResize: this.fitTextTransform,
          }, {
            default: () => avatatText,
          })
        }
        else if (this.$props.src) {
          return h('img', {
            destyler: 'avatar-img',
            ...this.$props.imgProps,
            loading:
            isImageSupportNativeLazy
            && !this.$props.intersectionObserverOptions
            && this.$props.lazy
              ? 'lazy'
              : 'eager',
            src:
            isImageSupportNativeLazy
              ? this.$props.src
              : this.shouldStartLoading || this.loaded
                ? this.$props.src
                : null,
            onLoad: this.mergedOnLoad,
            onError: this.handleError,
          })
        }
      })
    }
    return h('span', {
      ref: 'selfRef',
      destyler: 'avatar',
    }, [
      img,
      this.$props.lazy && placeholderNode,
    ])
  },
})

export {
  DestylerAvatar,
}
