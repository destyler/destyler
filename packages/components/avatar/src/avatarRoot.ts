import type { Ref } from 'vue'
import { defineComponent, h, ref } from 'vue'
import { DestylerPrimitive, destylerPrimitiveProp } from '@destyler/primitive'
import { createContext } from '@destyler/shared'
import type { ImageLoadingStatus } from '@destyler/composition'

interface AvatarRootContext {
  imageLoadingStatus: Ref<ImageLoadingStatus>
}

const [injectAvatarRootContext, provideAvatarRootContext] = createContext<AvatarRootContext>('DestylerAvatarRoot')

const DestylerAvatarRoot = defineComponent({
  name: 'DestylerAvatarRoot',
  props: destylerPrimitiveProp,
  setup() {
    provideAvatarRootContext({
      imageLoadingStatus: ref<ImageLoadingStatus>('loading'),
    })
  },
  render() {
    return h(DestylerPrimitive, {
      as: 'span',
      asChild: this.$props.asChild,
    }, this.$slots.default?.())
  },
})

export {
  DestylerAvatarRoot,
  AvatarRootContext,
  injectAvatarRootContext,
  provideAvatarRootContext,
}
