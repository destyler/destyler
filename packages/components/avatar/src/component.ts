export { DestylerAvatarFallback, type DestylerAvatarFallbackProps } from './avatarFallback'
export { DestylerAvatarImage, type DestylerAvatarImageProps } from './avatarImage'
export { DestylerAvatarRoot, type DestylerAvatarRootProps } from './avatarRoot'
