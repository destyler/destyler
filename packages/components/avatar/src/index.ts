export * from './avatarRoot'
export * from './avatarImage'
export * from './avatarFallback'
export * from './resolver'
