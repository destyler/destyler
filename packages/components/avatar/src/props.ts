export interface DestylerAvatarProps {
  src?: string
  alt?: string
  delayMs?: number
  fallback?: string
}
