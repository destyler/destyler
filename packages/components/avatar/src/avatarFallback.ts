import type { PropType } from 'vue'
import { defineComponent, h, ref, watch } from 'vue'
import { DestylerPrimitive } from '@destyler/primitive'
import type { ExtractPublicPropTypes } from '@destyler/shared'

import { injectAvatarRootContext } from './avatarRoot'

export const destylerAvatarFallbackProps = {
  asChild: {
    type: Boolean as PropType<boolean>,
    required: false,
    default: false,
  },
  delayMs: {
    type: Number as PropType<number>,
    required: false,
    default: 0,
  },
}

export type DestylerAvatarFallbackProps = ExtractPublicPropTypes<typeof destylerAvatarFallbackProps>

export const DestylerAvatarFallback = defineComponent({
  name: 'DestylerAvatarFallback',
  props: destylerAvatarFallbackProps,
  setup(props) {
    const rootContext = injectAvatarRootContext()

    const canRender = ref(false)
    let timeout: ReturnType<typeof setTimeout> | undefined

    watch(rootContext.imageLoadingStatus, (value) => {
      if (value === 'loading') {
        canRender.value = false
        if (props.delayMs) {
          timeout = setTimeout(() => {
            canRender.value = true
            clearTimeout(timeout)
          }, props.delayMs)
        }
        else {
          canRender.value = true
        }
      }
    }, { immediate: true })

    return {
      canRender,
      rootContext,
    }
  },
  render() {
    const useVShow = this.rootContext.imageLoadingStatus.value !== 'loaded' && this.canRender
    return useVShow
      ? h(DestylerPrimitive, {
        as: 'span',
        asChild: this.$props.asChild,
      }, this.$slots.default?.())
      : null
  },
})
