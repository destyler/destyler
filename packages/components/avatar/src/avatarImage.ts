import type { PropType } from 'vue'
import { defineComponent, h, toRefs, watch } from 'vue'
import { DestylerPrimitive, destylerPrimitiveProp } from '@destyler/primitive'
import { useImageLoadingStatus } from '@destyler/composition'
import { injectAvatarRootContext } from './avatarRoot'

const destylerAvatarImageProps = {
  ...destylerPrimitiveProp,
  src: {
    type: String as PropType<string>,
    required: true,
    default: '',
  },
}

const DestylerAvatarImage = defineComponent({
  name: 'DestylerAvatarImage',
  props: destylerAvatarImageProps,
  emits: ['loadingStatusChange'],
  setup(props, { emit }) {
    const { src } = toRefs(props)
    const rootContext = injectAvatarRootContext()

    const imageLoadingStatus = useImageLoadingStatus(src)

    watch(
      imageLoadingStatus,
      (newValue) => {
        emit('loadingStatusChange', newValue)
        if (newValue !== 'idle')
          rootContext.imageLoadingStatus.value = newValue
      },
      { immediate: true },
    )

    return {
      rootContext,
      src,
    }
  },
  render() {
    const useVShow = this.rootContext.imageLoadingStatus.value === 'loaded'
    return useVShow
      ? h(DestylerPrimitive, {
        as: 'img',
        role: 'img',
        asChild: this.$props.asChild,
        src: this.src,
      }, this.$slots.default?.())
      : null
  },
})

export {
  DestylerAvatarImage,
  destylerAvatarImageProps,
}
