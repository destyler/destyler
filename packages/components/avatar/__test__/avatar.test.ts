import { describe, expect, it, vi } from 'vitest'
import { mount } from '@vue/test-utils'
import { DestylerAvatar } from '../src'

describe('avatar', () => {
  it('image avatar', () => {
    const wrapper = mount(DestylerAvatar, {
      props: {
        src: 'https://github.com/elonehoo.png',
      },
    })
    expect(wrapper.find('img').exists()).toBe(true)
    expect(wrapper.html()).toMatchSnapshot()
    wrapper.unmount()
  })
  it('image avatar error handle when load failed', async () => {
    const onError = vi.fn()
    const wrapper = mount(DestylerAvatar, {
      props: {
        src: 'https://github.com/elonehoo.png',
        onError,
      },
    })
    await wrapper.find('img').trigger('error')
    expect(onError).toHaveBeenCalled()
    wrapper.unmount()
  })
})
