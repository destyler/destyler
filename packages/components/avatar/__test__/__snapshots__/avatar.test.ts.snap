// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Avatar > image avatar 1`] = `"<span destyler="avatar"><img destyler="avatar-img" src="https://github.com/elonehoo.png"><!----></span>"`;
